File: php/YurticiKargoClient.php
<?php
/**
 * Yurtiçi Kargo PHP API Client
 * 
 * Yurtiçi Kargo SOAP Web Service entegrasyonu için ana istemci sınıfı.
 * Tüm fonksiyonları birleştirir ve tek bir arayüz sunar.
 * 
 * @requires PHP 8.0+, ext-soap
 */

require_once __DIR__ . '/functions/CreateShipment.php';
require_once __DIR__ . '/functions/CancelShipment.php';
require_once __DIR__ . '/functions/QueryShipment.php';
require_once __DIR__ . '/functions/SaveReturnShipmentCode.php';

class YurticiKargoClient
{
    private const WSDL_TEST = 'https://testws.yurticikargo.com/KOPSWebServices/ShippingOrderDispatcherServices?wsdl';
    private const WSDL_LIVE = 'https://ws.yurticikargo.com/KOPSWebServices/ShippingOrderDispatcherServices?wsdl';

    private SoapClient $soapClient;
    private string $username;
    private string $password;
    private string $language;
    private bool $testMode;

    private CreateShipment $createShipmentService;
    private CancelShipment $cancelShipmentService;
    private QueryShipment $queryShipmentService;
    private SaveReturnShipmentCode $saveReturnShipmentCodeService;

    /**
     * @param string $username API kullanıcı adı
     * @param string $password API şifresi
     * @param string $language Dil (TR, EN)
     * @param bool $testMode Test ortamı mı?
     * @param array $soapOptions Ek SOAP istemci ayarları
     * @throws SoapFault
     */
    public function __construct(
        string $username,
        string $password,
        string $language = 'TR',
        bool $testMode = true,
        array $soapOptions = []
    ) {
        $this->username = $username;
        $this->password = $password;
        $this->language = $language;
        $this->testMode = $testMode;

        $wsdl = $testMode ? self::WSDL_TEST : self::WSDL_LIVE;

        $defaultOptions = [
            'trace' => true,
            'exceptions' => true,
            'encoding' => 'UTF-8',
            'soap_version' => SOAP_1_1,
            'connection_timeout' => 30,
            'location' => $testMode 
                ? 'https://testws.yurticikargo.com/KOPSWebServices/ShippingOrderDispatcherServices'
                : 'https://ws.yurticikargo.com/KOPSWebServices/ShippingOrderDispatcherServices',
            'stream_context' => stream_context_create([
                'ssl' => [
                    'verify_peer' => false,
                    'verify_peer_name' => false,
                    'allow_self_signed' => true,
                ]
            ]),
        ];

        $options = array_merge($defaultOptions, $soapOptions);

        $this->soapClient = new SoapClient($wsdl, $options);

        // Servisleri başlat
        $this->createShipmentService = new CreateShipment($this->soapClient, $this->username, $this->password, $this->language);
        $this->cancelShipmentService = new CancelShipment($this->soapClient, $this->username, $this->password, $this->language);
        $this->queryShipmentService = new QueryShipment($this->soapClient, $this->username, $this->password, $this->language);
        $this->saveReturnShipmentCodeService = new SaveReturnShipmentCode($this->soapClient, $this->username, $this->password, $this->language);
    }

    /**
     * Gönderi oluşturur.
     *
     * @param array $shipments Gönderi dizisi
     * @return object
     * @throws SoapFault|InvalidArgumentException
     */
    public function createShipment(array $shipments): object
    {
        return $this->createShipmentService->execute($shipments);
    }

    /**
     * Gönderileri iptal eder.
     *
     * @param array $cargoKeys İptal edilecek kargo anahtarları
     * @return object
     * @throws SoapFault|InvalidArgumentException
     */
    public function cancelShipment(array $cargoKeys): object
    {
        return $this->cancelShipmentService->execute($cargoKeys);
    }

    /**
     * Gönderi sorgulama yapar.
     *
     * @param array $keys Sorgulanacak anahtarlar
     * @param int $keyType 0=cargoKey, 1=invoiceKey
     * @param bool $addHistoricalData Hareket geçmişi dahil edilsin mi
     * @param bool $onlyTracking Sadece takip URL'si
     * @return object
     * @throws SoapFault|InvalidArgumentException
     */
    public function queryShipment(array $keys, int $keyType = 0, bool $addHistoricalData = false, bool $onlyTracking = false): object
    {
        return $this->queryShipmentService->execute($keys, $keyType, $addHistoricalData, $onlyTracking);
    }

    /**
     * İade kodu oluşturur.
     *
     * @param string $returnCode İade kodu
     * @param string $startDate Başlangıç tarihi (YYYYMMDD)
     * @param string $endDate Bitiş tarihi (YYYYMMDD)
     * @param int $maxCount Kullanım adedi
     * @param string|null $fieldName Özel alan (null ise test/canlı'ya göre otomatik atanır)
     * @return object
     * @throws SoapFault|InvalidArgumentException
     */
    public function saveReturnShipmentCode(string $returnCode, string $startDate, string $endDate, int $maxCount = 1, ?string $fieldName = null): object
    {
        if ($fieldName === null) {
            $fieldName = $this->testMode ? '53' : '16';
        }

        return $this->saveReturnShipmentCodeService->execute($returnCode, $startDate, $endDate, $maxCount, $fieldName);
    }

    // ─── Debug Metotları ───────────────────────────────────────────────────────

    /**
     * Son SOAP isteğini döndürür.
     */
    public function getLastRequest(): ?string
    {
        return $this->soapClient->__getLastRequest();
    }

    /**
     * Son SOAP yanıtını döndürür.
     */
    public function getLastResponse(): ?string
    {
        return $this->soapClient->__getLastResponse();
    }

    /**
     * SOAP servisinin desteklediği fonksiyonları listeler.
     */
    public function getFunctions(): array
    {
        return $this->soapClient->__getFunctions();
    }

    /**
     * SOAP servisinin desteklediği tipleri listeler.
     */
    public function getTypes(): array
    {
        return $this->soapClient->__getTypes();
    }

    /**
     * Test modunda mı çalışıyor?
     */
    public function isTestMode(): bool
    {
        return $this->testMode;
    }

    /**
     * Kullanılan WSDL URL'sini döndürür.
     */
    public function getWsdlUrl(): string
    {
        return $this->testMode ? self::WSDL_TEST : self::WSDL_LIVE;
    }
}

File: python/yurticikargo_client.py
"""
Yurtiçi Kargo Python API Client - Controller

Tüm API fonksiyonlarını tek bir sınıf altında birleştirir.
Sadece Python standard library kullanır (urllib, xml.etree, ssl, dataclasses).
"""

from __future__ import annotations

from typing import Any

from functions import (
    create_shipment as _create_shipment_fn,
    cancel_shipment as _cancel_shipment_fn,
    query_shipment as _query_shipment_fn,
    save_return_shipment_code as _save_return_code_fn,
)
from functions.create_shipment import ShipmentResult
from functions.cancel_shipment import CancelResult
from functions.query_shipment import QueryResult
from functions.save_return_shipment_code import ReturnCodeResult


class YurticiKargoClient:
    """
    Yurtiçi Kargo SOAP API istemcisi.

    Tüm API metodlarını method olarak sunar.
    Sadece Python standard library kullanır.

    Kullanım:
        client = YurticiKargoClient(username="YKTEST", password="YK", test_mode=True)
        result = client.create_shipment([{...}])
    """

    WSDL_TEST = "https://testws.yurticikargo.com/KOPSWebServices/ShippingOrderDispatcherServices?wsdl"
    WSDL_LIVE = "https://ws.yurticikargo.com/KOPSWebServices/ShippingOrderDispatcherServices?wsdl"

    def __init__(
        self,
        username: str = "YKTEST",
        password: str = "YK",
        language: str = "TR",
        test_mode: bool = True,
    ) -> None:
        """
        Args:
            username: WS kullanıcı adı
            password: WS şifresi
            language: Dil (TR/EN)
            test_mode: True ise test ortamı kullanılır
        """
        self.username = username
        self.password = password
        self.language = language
        self.test_mode = test_mode

    @property
    def wsdl_url(self) -> str:
        """Aktif WSDL URL'si."""
        return self.WSDL_TEST if self.test_mode else self.WSDL_LIVE

    def create_shipment(self, shipments: list[dict[str, Any]]) -> ShipmentResult:
        """
        Gönderi oluşturma.

        Args:
            shipments: Gönderi bilgileri listesi. Her dict en az şu alanları içermeli:
                - cargo_key (str): Benzersiz kargo anahtarı
                - invoice_key (str): Fatura anahtarı
                - receiver_cust_name (str): Alıcı adı
                - receiver_address (str): Alıcı adresi
                - receiver_phone1 (str): Alıcı telefon

        Returns:
            ShipmentResult: Sonuç bilgisi
        """
        return _create_shipment_fn(
            shipments=shipments,
            username=self.username,
            password=self.password,
            language=self.language,
            test_mode=self.test_mode,
        )

    def cancel_shipment(self, cargo_keys: list[str]) -> CancelResult:
        """
        Gönderi iptal etme.

        Args:
            cargo_keys: İptal edilecek kargo anahtarları

        Returns:
            CancelResult: Sonuç bilgisi
        """
        return _cancel_shipment_fn(
            cargo_keys=cargo_keys,
            username=self.username,
            password=self.password,
            language=self.language,
            test_mode=self.test_mode,
        )

    def query_shipment(
        self,
        keys: list[str],
        key_type: int = 0,
        add_historical_data: bool = False,
        only_tracking: bool = False,
    ) -> QueryResult:
        """
        Gönderi sorgulama.

        Args:
            keys: Sorgulanacak anahtar değerleri
            key_type: 0=Cargo Key, 1=Invoice Key
            add_historical_data: Hareket geçmişi dahil edilsin mi
            only_tracking: Sadece takip URL dönsün mü

        Returns:
            QueryResult: Sonuç bilgisi
        """
        return _query_shipment_fn(
            keys=keys,
            key_type=key_type,
            add_historical_data=add_historical_data,
            only_tracking=only_tracking,
            username=self.username,
            password=self.password,
            language=self.language,
            test_mode=self.test_mode,
        )

    def save_return_shipment_code(
        self,
        return_code: str,
        start_date: str,
        end_date: str,
        max_count: int = 1,
        field_name: str = "53",
    ) -> ReturnCodeResult:
        """
        İade kodu oluşturma.

        Args:
            return_code: İade kodu (benzersiz)
            start_date: Geçerlilik başlangıç tarihi (YYYYMMDD)
            end_date: Geçerlilik bitiş tarihi (YYYYMMDD)
            max_count: Kullanım adedi
            field_name: Özel alan bilgisi (test: '53' veya '3', canlı: '16')

        Returns:
            ReturnCodeResult: Sonuç bilgisi
        """
        return _save_return_code_fn(
            return_code=return_code,
            start_date=start_date,
            end_date=end_date,
            max_count=max_count,
            field_name=field_name,
            username=self.username,
            password=self.password,
            language=self.language,
            test_mode=self.test_mode,
        )

    def __repr__(self) -> str:
        mode = "TEST" if self.test_mode else "CANLI"
        return f"YurticiKargoClient(user={self.username!r}, mode={mode})"
